source/source_integration_test.go

```diff
@@ -57,13 +57,149 @@ func TestSource_SuccessfulSnapshot(t *testing.T) {
+	source := &Source{}
+
+	err := source.Configure(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = source.Open(ctx, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	numberOfFiles := 5
+	testFiles := addObjectsToTheBucket(ctx, t, testBucket, client, numberOfFiles)
+
+	// read and assert until last file is left
+	var record sdk.Record
+	for i := 0; i < numberOfFiles-1; i++ {
+		record, err = readAndAssert(ctx, t, source, testFiles[i])
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+	}
+
+	lastReadPosition := record.Position
+	if p, err := position.ParseRecordPosition(lastReadPosition); err != nil {
+		t.Fatal(err)
+	} else if p.Type != position.TypeSnapshot {
+		t.Fatalf("expected a position Type TypeSnapshot, got: %v", p.Type)
+	}
+
+	// Stop the source
+	_ = source.Teardown(ctx)
+
+	// Snapshot Restart will read all the files again
+
+	source = &Source{}
+	err = source.Configure(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Open with lastReadPosition
+	err = source.Open(ctx, lastReadPosition)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// read and assert all the files
+	for i := 0; i < numberOfFiles; i++ {
+		record, err = readAndAssert(ctx, t, source, testFiles[i])
 		if err != nil {
 			t.Fatalf("unexpected error: %v", err)
 		}
 	}
 
+	// As the last item reached now the poistion will be cdc
+	if p, err := position.ParseRecordPosition(record.Position); err != nil {
+		t.Fatal(err)
+	} else if p.Type != position.TypeCDC {
+		t.Fatalf("expected a position Type CDC, got: %v", p.Type)
+	}
+
+	_, err = source.Read(ctx)
+	if !errors.Is(err, sdk.ErrBackoffRetry) {
+		t.Fatalf("expected a BackoffRetry error, got: %v", err)
+	}
+}
+
+func TestSource_SnapshotRestartAfterLastRecord(t *testing.T) {
+	client, cfg := prepareIntegrationTest(t)
+
+	ctx := context.Background()
+	testBucket := cfg[config.ConfigKeyGCSBucket]
+	source := &Source{}
+
+	err := source.Configure(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = source.Open(ctx, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	numberOfFiles := 5
+	testFiles := addObjectsToTheBucket(ctx, t, testBucket, client, numberOfFiles)
+
+	// read and assert until last file is left
+	var record sdk.Record
+	for i := 0; i < numberOfFiles; i++ {
+		record, err = readAndAssert(ctx, t, source, testFiles[i])
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+	}
+
+	lastReadPosition := record.Position
+	if p, err := position.ParseRecordPosition(lastReadPosition); err != nil {
+		t.Fatal(err)
+	} else if p.Type != position.TypeCDC {
+		t.Fatalf("expected a position Type TypeCDC, got: %v", p.Type)
+	}
+
+	// Stop the source
+	_ = source.Teardown(ctx)
+
+	// Snapshot Restart will not read all the files again instead returns ErrBackoffRetry
+
+	source = &Source{}
+	err = source.Configure(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Open with lastReadPosition
+	err = source.Open(ctx, lastReadPosition)
+	if err != nil {
+		t.Fatal(err)
+	}
+
 	_, err = source.Read(ctx)
```

Comment:
assert the position type here to check if it's CDC plz

Reply:
so it returns ErrBackoffRetry because there is no updates right? maybe let's update something in the GCS here and make sure it's captured after the restart